lib/vtls/bearssl.c

```diff
@@ -277,35 +291,52 @@ static void x509_append(const br_x509_class **ctx, const unsigned char *buf,
 {
   struct x509_context *x509 = (struct x509_context *)ctx;
 
+  if(!x509->verifypeer) {
+    if(x509->cert_num == 0)
+      br_x509_decoder_push(&x509->decoder, buf, len);
+    return;
+  }
+
   x509->minimal.vtable->append(&x509->minimal.vtable, buf, len);
 }
 
 static void x509_end_cert(const br_x509_class **ctx)
 {
   struct x509_context *x509 = (struct x509_context *)ctx;
 
+  if(!x509->verifypeer) {
+    x509->cert_num++;
+    return;
+  }
+
   x509->minimal.vtable->end_cert(&x509->minimal.vtable);
 }
 
 static unsigned x509_end_chain(const br_x509_class **ctx)
 {
   struct x509_context *x509 = (struct x509_context *)ctx;
-  unsigned err;
 
-  err = x509->minimal.vtable->end_chain(&x509->minimal.vtable);
-  if(err && !x509->verifypeer) {
-    /* ignore any X.509 errors */
-    err = BR_ERR_OK;
+  if(!x509->verifypeer) {
+    return br_x509_decoder_last_error(&x509->decoder);
```

Comment:
Why did you change this to return the error in the decoder when !verifypeer?

Reply:
BearSSL has two APIs that deal with X.509 certificates. One is `br_x509_minimal`, which decodes a chain of certificates, verifying that they are valid, each one certifies the preceding one, and that the root certificate is in the trust anchor list. At the end of the chain, if there is no other error, it returns 0 if trusted, and `BR_ERR_X509_NOT_TRUSTED` if not. In either case, you can access the public key of the end-entity using the `get_pkey` function.

However, if there was some other error (for example, a DN mismatch, where a certificate doesn't certify the preceding one, or if one of the certificates is expired), then `br_x509_minimal` just stops processing the chain, and returns an error from `end_chain`. In this case, `get_pkey` will return NULL, even if it has decoded a key.

So, if we wish to ignore errors with expired or out-of-order certificates, then `br_x509_minimal` is not useful. Instead, we can use the second API, `br_x509_decoder`, which decodes a single certificate. This commit switches the curl x509 engine to using `br_x509_decoder` if !verifypeer, and `br_x509_minimal` otherwise.

The overall effect of this change is the following:
If verifypeer is false, curl will now *only* consider the first certificate in the chain sent by the server. If it can decode the first certificate and extract a key (regardless of the contents of subsequent certificates), then it accepts the chain. Previously, curl would accept the chain *only* if it was completely valid apart from none of the certificates being trusted.

Reply:
Ok thanks for the thorough explanation